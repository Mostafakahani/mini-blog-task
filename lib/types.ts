export interface Post {
  id: number;
  userId: number;
  title: string;
  body: string;
  user?: User;
}

export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  website: string;
}

export interface PostWithAuthor extends Post {
  author: string;
}
