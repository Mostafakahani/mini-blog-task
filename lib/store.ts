import { create } from "zustand";
import { persist } from "zustand/middleware";

export interface Post {
  id: number;
  title: string;
  body: string;
  userId: number;
}

interface PostState {
  posts: Post[];
  isLoading: boolean;
  error: string | null;
  fetchPosts: () => Promise<void>;
  getPostById: (id: number) => Post | undefined;
  addPost: (post: Omit<Post, "id">) => Promise<Post>;
}

export const usePostStore = create<PostState>()(
  persist(
    (set, get) => ({
      posts: [],
      isLoading: false,
      error: null,

      fetchPosts: async () => {
        set({ isLoading: true, error: null });
        try {
          const response = await fetch(
            `${process.env.NEXT_PUBLIC_BASE_URL || ""}/posts`
          );

          if (!response.ok) {
            throw new Error("Failed to fetch posts");
          }

          const data = await response.json();
          set({ posts: data, isLoading: false });
        } catch (error) {
          set({
            error:
              error instanceof Error ? error.message : "Unknown error occurred",
            isLoading: false,
          });
        }
      },

      getPostById: (id) => {
        return get().posts.find((post) => post.id === id);
      },

      addPost: async (newPost) => {
        set({ isLoading: true, error: null });
        try {
          const response = await fetch(
            `${process.env.NEXT_PUBLIC_BASE_URL || ""}/posts`,
            {
              method: "POST",
              headers: {
                "Content-Type": "application/json",
              },
              body: JSON.stringify(newPost),
            }
          );

          if (!response.ok) {
            throw new Error("Failed to create post");
          }

          const createdPost = await response.json();
          set((state) => ({
            posts: [...state.posts, createdPost],
            isLoading: false,
          }));

          return createdPost;
        } catch (error) {
          set({
            error:
              error instanceof Error ? error.message : "Unknown error occurred",
            isLoading: false,
          });
          throw error;
        }
      },
    }),
    {
      name: "posts-storage", // نام برای ذخیره در localStorage
    }
  )
);
