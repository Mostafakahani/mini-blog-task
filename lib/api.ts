import React from "react";
import { usePostStore } from "./store";
import { Post } from "./types";

const BASE_URL_POSTS = process.env.NEXT_PUBLIC_BASE_URL_POSTS || "";
const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL || "";

export async function getAllPosts() {
  try {
    // During build time, we should only fetch from external API
    // For server-side rendering that runs at request time, we can use both
    const isServerSideRendering =
      typeof window === "undefined" && process.env.NODE_ENV !== "production";

    if (isServerSideRendering) {
      // When in production build process, only fetch from external API
      const response = await fetch(`${BASE_URL_POSTS}/posts`);

      if (!response.ok) {
        throw new Error(`Failed to fetch posts from ${BASE_URL_POSTS}/posts`);
      }

      const data = await response.json();

      return {
        data: data.slice(0, 10),
        error: null,
      };
    } else {
      // In development or client-side, fetch from both sources
      try {
        const [response, responseLocal] = await Promise.all([
          fetch(`${BASE_URL_POSTS}/posts`),
          fetch(`${BASE_URL}/api/posts`),
        ]);

        if (!response.ok) {
          throw new Error(`Failed to fetch posts from ${BASE_URL_POSTS}/posts`);
        }

        const data = await response.json();
        const localData = await responseLocal.json();
        console.log({ data, localData });

        return {
          data: [...data.slice(0, 10), ...localData],
          localData,
          error: null,
        };
      } catch {
        // Fallback to just the external API if local API fails
        const response = await fetch(`${BASE_URL_POSTS}/posts`);

        if (!response.ok) {
          throw new Error(`Failed to fetch posts from ${BASE_URL_POSTS}/posts`);
        }

        const data = await response.json();

        console.warn(
          "Could not fetch from local API, using only external data"
        );
        return {
          data: data.slice(0, 10),
          error: null,
        };
      }
    }
  } catch (error) {
    console.error("Error fetching posts:", error);
    return {
      data: [],
      error: error instanceof Error ? error.message : "Unknown error occurred",
    };
  }
}
// Get post by ID
export async function getPostById(id: number) {
  try {
    const [response, responseLocal] = await Promise.all([
      fetch(`${BASE_URL_POSTS}/posts/${id}`),
      fetch(`${BASE_URL}/api/posts?id=${id}`),
    ]);

    if (!responseLocal.ok && !response.ok) {
      throw new Error("Failed to fetch post");
    }

    const data = responseLocal.ok
      ? await responseLocal.json()
      : await response.json();
    return { data, error: null };
  } catch (error) {
    return {
      data: {},
      error: error instanceof Error ? error.message : "Unknown error occurred",
    };
  }
}

// Get all users
export async function getAllUsers() {
  try {
    const response = await fetch(`${BASE_URL_POSTS}/users`);

    if (!response.ok) {
      throw new Error("Failed to fetch users");
    }

    const data = await response.json();
    return { data, error: null };
  } catch (error) {
    return {
      data: [],
      error: error instanceof Error ? error.message : "Unknown error occurred",
    };
  }
}

// Get user by ID
export async function getUserById(id: number) {
  try {
    const response = await fetch(`${BASE_URL_POSTS}/users/${id}`);

    if (!response.ok) {
      throw new Error("Failed to fetch user");
    }

    const data = await response.json();
    return { data, error: null };
  } catch (error) {
    return {
      data: {},
      error: error instanceof Error ? error.message : "Unknown error occurred",
    };
  }
}

// Create new post
export async function createPost(post: {
  title: string;
  body: string;
  user: object;
}): Promise<{ data?: Post; error?: string }> {
  try {
    const response = await fetch(`${BASE_URL}/api/posts`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(post),
    });

    if (!response.ok) {
      throw new Error("خطا در ایجاد پست");
    }

    const data = await response.json();
    return { data, error: undefined };
  } catch (error) {
    return {
      // data: {},
      error:
        error instanceof Error ? error.message : "خطای ناشناخته رخ داده است",
    };
  }
}

// Update post
export async function updatePost(
  id: number,
  post: {
    title: string;
    body: string;
    user?: object;
  }
): Promise<{ data?: Post; error?: string }> {
  try {
    const response = await fetch(`${BASE_URL}/api/posts?id=${id}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(post),
    });

    if (!response.ok) {
      throw new Error("خطا در بروزرسانی پست");
    }

    const data = await response.json();
    return { data, error: undefined };
  } catch (error) {
    return {
      // data: {},
      error:
        error instanceof Error ? error.message : "خطای ناشناخته رخ داده است",
    };
  }
}

// Get all posts with Zustand
export function usePosts() {
  const { posts, isLoading, error, fetchPosts } = usePostStore();

  React.useEffect(() => {
    if (posts.length === 0 && !isLoading && !error) {
      fetchPosts();
    }
  }, [posts.length, isLoading, error, fetchPosts]);

  return { posts, isLoading, error, refetch: fetchPosts };
}

// Get post by ID with Zustand
export function usePost(id: number) {
  const { posts, getPostById, isLoading, error, fetchPosts } = usePostStore();

  React.useEffect(() => {
    if (posts.length === 0 && !isLoading && !error) {
      fetchPosts();
    }
  }, [posts.length, isLoading, error, fetchPosts]);

  const post = getPostById(id);
  return { post, isLoading, error };
}
