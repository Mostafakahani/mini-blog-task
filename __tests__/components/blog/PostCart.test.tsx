import { render, screen } from "@testing-library/react";
import { Post, PostWithAuthor } from "@/lib/types";
import PostCard from "@/components/blog/PostCart";

const longPost: PostWithAuthor = {
  id: 1,
  author: "John Doe",
  userId: 101,
  title: "Test Post Title",
  body: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum ac diam sit amet quam vehicula elementum sed sit amet dui. Donec sollicitudin, magna ut pretium.",
};

const shortPost: PostWithAuthor = {
  id: 2,
  userId: 102,
  title: "Short Post",
  body: "Short body text.",
  author: "Jane Smith",
};

describe("PostCard", () => {
  it("renders the post title and truncated body when body is longer than 100 characters", () => {
    render(<PostCard post={longPost} />);
    expect(screen.getByText(longPost.title)).toBeInTheDocument();
    const truncatedBody =
      longPost.body.length > 100
        ? longPost.body.substring(0, 100) + "..."
        : longPost.body;
    expect(screen.getByText(truncatedBody)).toBeInTheDocument();
  });

  it("renders the full body when body is short", () => {
    render(<PostCard post={shortPost} />);
    expect(screen.getByText(shortPost.title)).toBeInTheDocument();
    expect(screen.getByText(shortPost.body)).toBeInTheDocument();
  });

  it("renders a link with the correct href to the blog post", () => {
    render(<PostCard post={longPost} />);
    const linkElement = screen.getByRole("link", { name: /بیشتر بخوانید/i });
    expect(linkElement).toHaveAttribute("href", `/blog/${longPost.id}`);
  });

  it("renders the author name if post has an author", () => {
    render(<PostCard post={longPost} />);
    expect(screen.getByText(longPost.author)).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<PostCard post={longPost} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
