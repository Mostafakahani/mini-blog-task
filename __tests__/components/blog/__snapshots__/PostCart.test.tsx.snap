// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard matches snapshot 1`] = `
<div
  class="bg-white rounded-lg shadow-md overflow-hidden hover:shadow-lg transition-shadow h-full flex flex-col min-h-[350px] sm:min-h-[400px]"
>
  <div
    class="flex-1 flex flex-col"
  >
    <div
      class="relative w-full h-48 sm:h-56"
    >
      <img
        alt="Test Post Title"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2Fid%2F237%2F536%2F354&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="p-4 sm:p-6 flex-1 flex flex-col justify-between"
    >
      <h3
        class="text-lg sm:text-xl font-semibold text-gray-900 mb-3 line-clamp-2 font-lobester"
      >
        Test Post Title
      </h3>
      <p
        class="text-gray-600 mb-4 line-clamp-3 text-sm sm:text-base"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum ac diam sit amet quam vehicula e...
      </p>
      <div
        class="flex flex-row items-center justify-between w-full"
      >
        <p
          class="text-sm sm:text-sm text-gray-900/50 line-clamp-2 font-lobester"
        >
          John Doe
        </p>
        <a
          class="mt-auto inline-flex items-center text-blue-600 hover:text-blue-800 transition-colors text-sm sm:text-base font-medium px-2 py-1 group"
          href="/blog/1"
        >
          بیشتر بخوانید
          <svg
            class="h-4 w-4 mr-2 group-hover:-translate-x-1 transition-transform rotate-180"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 5l7 7-7 7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
</div>
`;
